File: Single_Number_II.java
/**
 * Problem: https://leetcode.com/problems/single-number-ii/
 * x, y 两个变量。
 * 当同一个数第一次出现的时候，x 被设为那个数。
 * 当第二次出现的时候，x被清0，y被设置为那个数。
 * 当第三次出现的时候，y被清0。
 * 因为有一个数只出现一次，所以最后x就是那个数.
 */

public class Solution {
    public int singleNumber(int[] nums) {
        int x = 0x0;
        int y = 0x0;
        for(int n : nums) {
            x ^= ~y & n;
            y ^= ~x & n;
        }
        return x;
    }
}

File: Swift/single-number-ii.swift
/**
 * https://leetcode.com/problems/single-number-ii/
 * 
 * 
 */ 

class Solution {
    func singleNumber(_ nums: [Int]) -> Int {
        return self.singleNumber(nums, with: 3)
    }
    
    fileprivate func singleNumber(_ nums: [Int], with repeatNumber: Int) -> Int {
        var ret = 0
        for bit in 0 ..< 64 {
            var count = 0
            for n in nums {
                if ((n >> bit) & 1) == 1 { count += 1 }
            }
            count %= repeatNumber
            ret |= (count << bit)
        }
        return ret
    }
    
    func singleNumber(_ nums: [Int]) -> Int {
        var count: [Int : Int] = [:]
        for n in nums {
            count[n] = (count[n, default: 0] + 1)
        }
        return count.filter{$0.value == 1}.keys.first ?? 0
    }
}
